### ISSUE TITLE
Quorum-based replication configuration options

### ISSUE THREAD

**Issue by sergos** [tarantool/tarantool#5073]
Qsync requires a number of options to configure the cluster.
Every space needs an extra field `is_sync` that can be set during the space creation:
`
box.schema.create_space('test', {is_sync = true})
`

Leader has no option but to support the synchronous replication of the spaces, so no need for extra option for the leader. As for the replicas - they have to follow the TXN+Confirm model, so no options too.
 
The only option is for the leader to consider a replica as a synchronous - means, take it's responses into account during quorum collection - and this can be reflected in the cluster configuration.
`
[1, '416ded45-52cb-482f-a440-d0e5518429a3', 'Leader']
`
`
[2, 'c32532b5-4a6e-4420-a058-604281331bd3']
`
`
[3, '5babcd4b-2eaa-4272-ab6f-ae4ac289f122', 'Sync']
`

The qsync should have its own timeouts, which are
`
box.cfg.synchro_replication_heartbeat = 4
`
`
box.cfg.synchro_replication_quorum_timeout = 4
`

Also cluster should be configured with quorum for the replication:
`
box.cfg.synchro_replication_quorum = 4
`
The quorum value can be from 0 to the number of synchronous nodes (including leader). Recommended value is number of synchronous nodes divided by 2 plus one. 

Added 07.08.2020

To enable the so-called asynchronous replicas - actually, replicas which do not participate in quorum collection - we need to implement a filter at the limbo votes collector. I propose to have a 32-bit mask that denotes with '1' replica that participate the election. This will solve the currently present problem with anonymous replicas participation, which means the mask should always be even.

The mask should be set on the leader node, the interface can be a direct update to the 32-bit integer with a checkup that the bit 0 is set to 0. The default value should be 0, otherwise addition of a new replica can cause problems with data safety, since this new replica can be intended for a temporary use
```
box.cfg.synchro_replication_quorum_mask = 0
```
I heard a number of complains the bitset is not readable for a 'regular user' so it should be also dubbed with
```
box.cgf.synchro_replication_quorum_mask_set(ID, [1|0])
```
Then user can easily add 3rd replica with ```box.cgf.synchro_replication_quorum_mask_set(3)``` to the quorum participants, or remove it with ```box.cgf.synchro_replication_quorum_mask_set(3, 0)```.

The mask have to be replicated through the _cluster space so that change of it on a leader will be delivered to all cluster participants.

**Comment by Gerold103** [comment]
Not counting option names being inconsistent with other replication options, what I don't understand the most is
- What is that heartbeat thing? We already have heartbeats;
- What is 'Leader', 'Sync' and other replicas? We don't divide replicas on sync and async. Only transactions can be sync or async. If you say, that a replica is sync, it means all transaction should be sync and be committed only after this replica receives them. When we talk about sync transactions, they don't care to which replicas they managed to replicate - the only important parameter is the quorum.

**Comment by sergos** [comment]
True, we can reuse the regular heartbeat.
As for division of replicas: we should have an explicit way to configure which of our replicas should participate in quorum and which are not. This will let ensure some DC will not appear with async-only replicas for example.

And let's rename the options for consistency into:

* replication_synchro_quorum_timeout = 4
* replication_synchro_quorum = 4

**Comment by Gerold103** [comment]
Well, in that case I don't understand the purpose of 'quorum'. If anyway you need to specify which replicas are sync and which are not. If you specify sync replicas, the quorum option is redundant, because it stops making any sense. Indeed, a sync transaction should be replicated to at least all these sync replicas. So quorum should be >= sync replica count. On the other hand, waiting for replication on the async replicas makes no sense either, because leads to the same problem - some DCs may receive more 'sync's. So having quorum > sync replica count is not needed, and it can be only equal. Then what is a purpose of `replication_synchro_quorum `?

**Comment by sergos** [comment]
The idea was to be tolerant for replicas failure. Consider you have 5 nodes, a leader and 4 replicas. All of them have to be sync ones, so leader waits for confirmation from each of 4. Still, leader can answer Ok after it obtains 2 confirms (which means we have quorum == 3, i.e. 5/2+1). 
Now, if one of these 4 replicas fails, cluster doesn't even feel it: we still have 3 replicas available to obtain 2 confirms without any action.

Cluster can have more async replicas beyond those 4 - they can be behind slow network, big distance or whatever else limitation to be synch ones. Each of async replica can be switched into sync mode, but it should be done as a separate action.

**Comment by Gerold103** [comment]
`Consider you have 5 nodes, a leader and 4 replicas. All of them have to be sync ones, so leader waits for confirmation from each of 4.`
This basically means the quorum is 5. A leader can't answer Ok, if quorum is not achieved, i.e. ack count < 5 (including local WAL). The whole RFC was based on quorum concept, and it is a real surprise to see the new concept of 'sync' replica now in the end of the design. I still don't understand why do we need both quorum option and 'sync' replicas is not answered.

**Comment by sergos** [comment]
> `Consider you have 5 nodes, a leader and 4 replicas. All of them have to be sync ones, so leader waits for confirmation from each of 4.`
> This basically means the quorum is 5. A leader can't answer Ok, if quorum is not achieved, i.e. ack count < 5 (including local WAL). The whole RFC was based on quorum concept, and it is a real surprise to see the new concept of 'sync' replica now in the end of the design. I still don't understand why do we need both quorum option and 'sync' replicas.

So, lets those 5 are ID 1..5. So, by introduction of a 32-bit ID mask of 0x0000003e you can annotate that only answers from those 5 are accounted for quorum. 

Now you can add nodes with ID 6+ in the same way as it is done now, so these replicas can be used in the cluster for some experiments, although will not affect the stability (by stealing votes from first five). Then the process of adding a voting replica will include an update of the mask. By default the mask can be 0xfffffffe to support current behavior - all replicas are voting.

The reasoning for regular (non-anonymous) replicas was described multiple times by @Mons, although I can find no documents about it. I hope he can add a brief comment here.

**Comment by Mons** [comment]
1. I consider useless separate option `box.cfg.synchro_replication_quorum` ,since it should be counted from members of cluster as `N/2+1`
2. I insist on separate roles on replicas (sync and common), so we could build cluster with reasonably small amount of synchronous replicas. And quorum will be achieved by counting N as sum(sync replicas) + leader

**Comment by sergos** [comment]
1) contradicts the 2) actually. Consider a cluster of 13 nodes with only 3 voting (including leader). Means you'll never achieve the ```N/2+1==7``` quorum.
On the other hand, one can have a cluster of 3 geo locations, 3 nodes at each location. The quorum is 5, means one of location can be out of votes completely and we're getting into an uber-complicated config with [geo] groups.

And - there were no justification for the async versus anonymous. Could you please, @Mons?

**Comment by Mons** [comment]
> Consider a cluster of 13 nodes with only 3 voting (including leader)... `N/2+1==7`
You're wrong for this case. N = 3, not 13. 

Anon replica is replica with id=0. It could ever became maser under no circumstances. By design.
Async replica is a common replica with id > 0, which could potentially became a master.

BTW: what about interrelation between master-master replication and q-based replication?
These days we have no "replicas": we have only masters, some of which are set to read_only.

**Comment by sergos** [comment]
@Gerold103 I put the plan of implementation of the option to mask non-voting replicas.
@Mons how would you like to see it in box.info.replication? I propose to:
```
tarantool> box.info.replication
---
- 1:
    id: 1
    quorum: yes
    uuid: 52247308-f795-4e10-8ad4-d9a052413f83
    lsn: 470
  2:
    id: 2
    quorum: no
    uuid: 1adc8626-7a47-40f3-bc9b-5262b1a72517
    lsn: 0
    downstream:
      status: stopped
      message: unexpected EOF when reading from socket, called on fd 15, aka 127.0.0.1:3333
      system_message: Broken pipe
  3:
    id: 3
    quorum: yes
    uuid: 45b032f8-7fa8-4a24-b1e8-3328b317f463
    lsn: 0
    downstream:
      status: follow
      idle: 0.61348325759172
      vclock: {1: 470}
...
```

**Comment by Gerold103** [comment]
@sergepetrenko is this ticket still making sense?

**Comment by sergepetrenko** [comment]
@Gerold103, no, looks mostly outdated.
We have another ticket for "learner" role: https://github.com/tarantool/tarantool/issues/6281, which's quite similar to this ticket's notion of `sync` and `async` replicas.
